api/types/resource.go

```diff
@@ -568,6 +559,22 @@ Outer:
 	return true
 }
 
+func containsFold(s, substr string) bool {
```

Comment:
nit: this helper is a bit funny looking and could probably use explanation.
```suggestion
// containsFold is a case-insensitive alternative to strings.Contains, used to help avoid excess allocations during searches.
func containsFold(s, substr string) bool {
```